components/TextFields/tests/unit/ContainedInputView/MDCBaseTextFieldTests.m

```diff
@@ -14,89 +14,527 @@
+  BOOL leadingViewVisibleBeforeBecomingFirstResponder = !textField.leadingView.hidden;
+  [self setIsEditing:YES onTextField:textField];
+  [self forceLayoutOfView:textField];
+  CGRect leadingFrameEditing = textField.leftView.frame;
+  BOOL leadingViewVisibleAfterBecomingFirstResponder = !textField.leadingView.hidden;
+
+  // Then
+  XCTAssertFalse(leadingViewVisibleBeforeBecomingFirstResponder);
+  XCTAssertTrue(leadingViewVisibleAfterBecomingFirstResponder);
+  XCTAssertFalse(CGRectEqualToRect(leadingFrameNotEditing, leadingFrameEditing));
+}
+
+- (void)testOnlyTrailingViewWithViewModeWhileEditing {
+  // Given
+  MDCBaseTextField *textField =
+      [self createLTRTextFieldWithLeadingView:NO
+                                 trailingView:YES
+                                     viewMode:UITextFieldViewModeWhileEditing];
+
+  // When
+  [self forceLayoutOfView:textField];
+  CGRect trailingFrameNotEditing = textField.rightView.frame;
+  BOOL trailingViewVisibleBeforeBecomingFirstResponder = !textField.trailingView.hidden;
+  [self setIsEditing:YES onTextField:textField];
+  [self forceLayoutOfView:textField];
+  BOOL trailingViewVisibleAfterBecomingFirstResponder = !textField.trailingView.hidden;
+  CGRect trailingFrameEditing = textField.rightView.frame;
+
+  // Then
+  XCTAssertFalse(trailingViewVisibleBeforeBecomingFirstResponder);
+  XCTAssertTrue(trailingViewVisibleAfterBecomingFirstResponder);
+  XCTAssertFalse(CGRectEqualToRect(trailingFrameNotEditing, trailingFrameEditing));
+}
+
+- (void)testLeadingAndTrailingViewsWithViewModeWhileEditing {
+  // Given
+  MDCBaseTextField *textField =
+      [self createLTRTextFieldWithLeadingView:YES
+                                 trailingView:YES
+                                     viewMode:UITextFieldViewModeWhileEditing];
+
+  // When
+  [self forceLayoutOfView:textField];
+  CGRect leadingFrameNotEditing = textField.leftView.frame;
+  CGRect trailingFrameNotEditing = textField.rightView.frame;
+  BOOL leadingViewVisibleBeforeBecomingFirstResponder = !textField.leadingView.hidden;
+  BOOL trailingViewVisibleBeforeBecomingFirstResponder = !textField.trailingView.hidden;
+  [self setIsEditing:YES onTextField:textField];
+  [self forceLayoutOfView:textField];
+  BOOL leadingViewVisibleAfterBecomingFirstResponder = !textField.leadingView.hidden;
+  BOOL trailingViewVisibleAfterBecomingFirstResponder = !textField.trailingView.hidden;
+  CGRect leadingFrameEditing = textField.leftView.frame;
+  CGRect trailingFrameEditing = textField.rightView.frame;
+
+  // Then
+  XCTAssertFalse(leadingViewVisibleBeforeBecomingFirstResponder);
+  XCTAssertFalse(trailingViewVisibleBeforeBecomingFirstResponder);
+  XCTAssertTrue(leadingViewVisibleAfterBecomingFirstResponder);
+  XCTAssertTrue(trailingViewVisibleAfterBecomingFirstResponder);
+  XCTAssertFalse(CGRectEqualToRect(leadingFrameNotEditing, leadingFrameEditing));
+  XCTAssertFalse(CGRectEqualToRect(trailingFrameNotEditing, trailingFrameEditing));
```

Comment:
General question: is it generally true that the leading/trailing frames will change when visibility changes? I'd expect only the views' hidden property to be toggled; is it possible for us to make the text field do so? Adjusting the frame of the left/right views could cause costly operations to occur (e.g. redrawing an icon), so if it's possible to easily avoid excessive layouts then that would be preferable.

Reply:
Not sure how it works with UITextField. I should look. With mine the way it's worked so far is their origins are made to be {0, 0}, but the size doesn't change, and they're hidden. This isn't because I felt it was the best way to do it necessarily, it was just one way that seemed to work so didn't think about it anymore.

Reply:
Ah; can we simplify it to just toggle the hidden property then?

Reply:
I only bother calculating origins of the side views if they're going to be displayed--that's why they're { 0, 0 } otherwise. It's actually simpler if I keep it the way it is. Is that alright?

Reply:
Extra paths of logic add complexity to a code base, and if you can accomplish the same thing with less code then that's usually preferable. I'm not seeing a strong case for why we need to conditionally update the layout of the views based on their visibility, so in this case the simpler logical path will be to simply update the visibility.

Please do so, barring a strong technical reason why it's beneficial to go with your current approach.

Reply:
Here's some pseudocode that shows how the left view frame is calculated:
```
if (shouldShowLeftView) {
    // Do the work of calculating what the origin should be so the frame looks okay
} else {
    // Don't bother, because nobody will see it
} 
```
If you want these invisible views to have certain frames at the cost of a little more math i'm more than happy to do that for you

Reply:
Yes, please remove that conditional path. We can add it or something like it back in the future if we determine that it adversely affects performance.